### PR TITLE
buildDotnetModule: support ReadyToRun and native binaries inside nuget files

### LINKED ISSUE #181373: dotnet: using nuget packages with native binaries in them
## Issue description

Some NuGet packages include native binaries inside of them for various purposes - ie. see [Grpc.Tools](https://www.nuget.org/packages/Grpc.Tools).

Using them in NixOS is currently impossible, since these binaries are compiled for a normal FHS linux system, and fail on NixOS.

Two additional features are needed for usage of such binaries to improve in NixOS:

1. Support in `buildDotnetModule`

   `buildDotnetModule` should automatically run `patchelf` on all found binaries in 
    the restored nuget dependencies. This would allow easy support in `nixpkgs`

2. Support when running bare `dotnet` command

    This one is a lot more trickier. When running native `dotnet` command (outside the 
    scope of a Nix build), `patchelf` should automatically patch restored nuget packages.

     This is a lower priority feature, mainly aimed at developer egonomics. It would 
     be also a lot harder to implement. Ideas welcome.

### Steps to reproduce

Try to create and build any .NET project that uses nuget packages with native binaries
(ie. `eventstore`)

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

### LINKED ISSUE #230848: `buildDotnetModule` with `ReadyToRun` restricts the derivation to a single platform
### Describe the bug

Since .NET6, the `crossgen2` used when using R2R, has been separated from the runtime. [[1]](https://learn.microsoft.com/en-us/dotnet/core/compatibility/sdk/6.0/publish-readytorun-requires-restore-change)
Therefore, when R2R is specified, `nix-build -A ${pkg}.fetch-deps` gets the platform-specific crossgen2. [[2]](https://github.com/NixOS/nixpkgs/pull/229434#discussion_r1182326011)
At this time, crossgen2 for multiple platforms cannot coexist, so the platform of derivation is restricted.

### Steps To Reproduce
1. Run `nix-build -A tone.passthru.fetch-deps` on platform other than "x86_64-linux". 

### Expected behavior
The package can be built on multiple platforms.

### Additional context
Disabling "R2R" solves this problem since the platform-dependent `crossgen2` is no longer used.
In this case, we can set it as the default value for `dotnetRestoreFlags` and `dotnetBuildFlags` (or `dotnetFlags`?).

### Notify maintainers
@IvarWithoutBones, @mdarocha, @corngood
I mentioned those who have committed to `buildDotnetModule` a lot.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.90.1-microsoft-standard-WSL2, NixOS, 23.05 (Stoat), 23.05.20230507.7975a3c`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.3`
 - channels(root): `"nixos-22.05"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by mdarocha**
###### Description of changes

Closes https://github.com/NixOS/nixpkgs/issues/230848

Closes https://github.com/NixOS/nixpkgs/issues/181373

This adds `crossgen2` nugets to the per-sdk nuget source, allowing them to be found
by the dotnet build process. 

Also, we iterate over the entire `.nuget` folder after restore, and perform `patchelf` on any native binaries found, allowing them to run properly during the build process. This allows dotnet packages that use `crossgen2` (`PublishReadyToRun` flag), and other nugets using native binaries, like `protoc` or `minver`, to work properly without any hacks.

This PR also contains small changes to how the `--runtime` flag is set, so that hopefully some errors can be avoided automatically (ie. running `PublishReadyToRun` may require the `--runtime` flag). The flag is now set when building a `.csproj` file or when `selfContainedBuild` is on (as before).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/dotnet/build-dotnet-module/hooks/dotnet-configure-hook.sh [hunk 1]
```diff
@@ -41,10 +39,33 @@ EOF
...
+
     for project in ${projectFile[@]} ${testProjectFile[@]-}; do
         dotnetRestore "$project"
     done
 
+    echo "Fixing up native binaries..."
+    # Find all native binaries and nuget libraries, and fix them up,
+    # by setting the proper interpreter and rpath to some commonly used libraries
+    for binary in $(find "$HOME/.nuget/packages/" -type f -executable); do
+        if patchelf --print-interpreter "$binary" >/dev/null 2>/dev/null; then
```
Comment (SuperSandro2000) [inline pkgs/build-support/dotnet/build-dotnet-module/hooks/dotnet-configure-hook.sh:52]:
```suggestion
        if patchelf --print-interpreter "$binary" &>/dev/null; then
```

File: pkgs/build-support/dotnet/build-dotnet-module/hooks/dotnet-configure-hook.sh [hunk 2]
```diff
@@ -41,10 +39,33 @@ EOF
...
+            # This makes sure that if the binary requires some specific runtime dependencies, it can find it.
+            # This fixes dotnet-built binaries like crossgen2
+            patchelf \
+                --add-needed libicui18n.so \
+                --add-needed libicuuc.so \
+                --add-needed libz.so \
+                --add-needed libssl.so \
+                "$binary"
+
+            patchelf --set-rpath "@libPath@" "$binary"
```
Comment (SuperSandro2000) [inline pkgs/build-support/dotnet/build-dotnet-module/hooks/dotnet-configure-hook.sh:65]: Can we combine those 3 commands into one?

**Review by raphaelr — COMMENTED** [review]
dotnet 8.0.nix needs to be regenerated as well (the current preview4 verison is in nixpkgs master now)

Otherwise LGTM

**Comment by mdarocha** [comment]
Updated, should be ready to merge

**Comment by mdarocha** [comment]
35 packages built:
ArchiSteamFarm BeatSaberModManager alttpr-opentracker audiobookshelf boogie btcpayserver btcpayserver-altcoins dafny denaro depotdownloader discordchatexporter-cli eventstore formula galaxy-buds-client inklecate jackett jellyfin kavita marksman msbuild nbxplorer netcoredbg omnisharp-roslyn opentabletdriver osu-lazer pinta ps3-disc-dumper roslyn ryujinx scarab slskd space-station-14-launcher tone wasabibackend xivlauncher

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>35 packages built:</summary>
  <ul>
    <li>ArchiSteamFarm</li>
    <li>BeatSaberModManager</li>
    <li>alttpr-opentracker</li>
    <li>audiobookshelf</li>
    <li>boogie</li>
    <li>btcpayserver</li>
    <li>btcpayserver-altcoins</li>
    <li>dafny</li>
    <li>denaro</li>
    <li>depotdownloader</li>
    <li>discordchatexporter-cli</li>
    <li>eventstore</li>
    <li>formula</li>
    <li>galaxy-buds-client</li>
    <li>inklecate</li>
    <li>jackett</li>
    <li>jellyfin</li>
    <li>kavita</li>
    <li>marksman</li>
    <li>msbuild</li>
    <li>nbxplorer</li>
    <li>netcoredbg</li>
    <li>omnisharp-roslyn</li>
    <li>opentabletdriver</li>
    <li>osu-lazer</li>
    <li>pinta</li>
    <li>ps3-disc-dumper</li>
    <li>roslyn</li>
    <li>ryujinx</li>
    <li>scarab</li>
    <li>slskd</li>
    <li>space-station-14-launcher</li>
    <li>tone</li>
    <li>wasabibackend</li>
    <li>xivlauncher</li>
  </ul>
</details>

**Review by raphaelr — APPROVED** [review]

**Comment by natsukium** [comment]
I confirmed that R2R works fine.
Thanks!

**Review by corngood — APPROVED** [review]

**Pull request merged**